File: c++/0905.cpp
// Link: https://leetcode.com/problems/sort-array-by-parity/
// The first solution
// Runtime: 91 ms
// Memory: 16.6 MB

class Solution
{
public:
    std::vector<int> sortArrayByParity(const std::vector<int>& nums)
    {
        std::vector<int> result;
        
        for (int value : nums)
            if (value % 2 == 0) result.insert(result.begin(), value);
            else result.insert(result.end(), value);
        
        return result;
    }
};

// The second solution
// Runtime: 16 ms
// Memory: 16.1 MB

class Solution
{
public:
    std::vector<int> sortArrayByParity(std::vector<int>& nums)
    {
        int left = 0, right = nums.size() - 1;
        while (left < right)
        {
            if (nums[left] % 2 != 0)
            {
                std::swap(nums[left], nums[right]);
                right--;
            }
            else left++;
        }
        
        return nums;
    }
};

// The third solution
// Runtime: 3 ms
// Memory: 16.56 MB

class Solution 
{
public:
    std::vector<int> sortArrayByParity(const std::vector<int>& nums) 
    {
        std::vector<int> sortedNums = nums;
        int left = 0;
        for (int right = 0; right < static_cast<int>(nums.size()); right++)
        {
            if (sortedNums[right] % 2 == 0) 
            {
                std::swap(sortedNums[left], sortedNums[right]);
                left++;
            }
        }

        return sortedNums;
    }
};

File: python/0905.py
# Link: https://leetcode.com/problems/sort-array-by-parity/description/
# The first solution
# Runtime: 68 ms
# Memory: 17.13 MB


class Solution:
    def sortArrayByParity(self, nums: List[int]) -> List[int]:
        sorted_nums = nums[:]
        left = 0
        for right in range(len(sorted_nums)):
            if sorted_nums[right] % 2 == 0:
                sorted_nums[left], sorted_nums[right] = (
                    sorted_nums[right],
                    sorted_nums[left],
                )
                left += 1

        return sorted_nums


# The second solution
# Runtime: 81 ms
# Memory: 17.05 MB


class Solution:
    def sortArrayByParity(self, nums: List[int]) -> List[int]:
        return [x for x in nums if x % 2 == 0] + [x for x in nums if x % 2 != 0]
